src/api/routes.rs

```diff
@@ -0,0 +1,876 @@
+            .find_one(
+                doc! {
+                    "inclusion_state": LedgerInclusionState::Included as u8 as i32,
+                    "message.payload.transaction_id": &transaction_id.to_string(),
+                },
+                None,
+            )
+            .await?
+            .ok_or_else(|| ListenerError::NoResults)?,
+    )?;
+
+    Ok(ListenerResponse::Message {
+        network_id: match &rec.message {
+            Message::Chrysalis(m) => Some(m.network_id()),
+            Message::Stardust(_) => None,
+        },
+        protocol_version: match &rec.message {
+            Message::Chrysalis(_) => 0,
+            Message::Stardust(m) => m.protocol_version(),
+        },
+        parents: rec.parents().map(|m| m.to_string()).collect(),
+        payload: match &rec.message {
+            Message::Chrysalis(m) => m.payload().as_ref().map(serde_json::to_value),
+            Message::Stardust(m) => m.payload().map(serde_json::to_value),
+        }
+        .transpose()
+        .map_err(|e| ListenerError::Other(e.into()))?,
+        nonce: rec.nonce(),
+    })
+}
+
+async fn milestone(database: Extension<Database>, Path((_ver, index)): Path<(APIVersion, u32)>) -> ListenerResult {
+    database
+        .collection::<Document>("messages")
+        .find_one(doc! {"message.payload.essence.index": &index}, None)
+        .await?
+        .ok_or_else(|| ListenerError::NoResults)
+        .and_then(|d| {
+            let rec = MessageRecord::try_from(d)?;
+            Ok(ListenerResponse::Milestone {
+                milestone_index: index,
+                message_id: rec.message_id.to_string(),
+                timestamp: match &rec.message {
+                    Message::Chrysalis(m) => {
+                        if let Some(crate::cpt2::payload::Payload::Milestone(m)) = m.payload() {
+                            m.essence().timestamp()
+                        } else {
+                            unreachable!()
+                        }
+                    }
+                    Message::Stardust(m) => {
+                        if let Some(crate::stardust::payload::Payload::Milestone(m)) = m.payload() {
+                            m.essence().timestamp()
+                        } else {
+                            unreachable!()
+                        }
+                    }
+                },
+            })
+        })
+}
+
+async fn address_analytics(
+    database: Extension<Database>,
+    TimeRange {
+        start_timestamp,
+        end_timestamp,
+    }: TimeRange,
+) -> ListenerResult {
+    let start_milestone = start_milestone(&database, start_timestamp).await?;
+    let end_milestone = end_milestone(&database, end_timestamp).await?;
+
+    let res = database
+        .collection::<Document>("messages")
+        .aggregate(
+            vec![
+                doc! { "$match": {
+                    "inclusion_state": LedgerInclusionState::Included as u8 as i32,
```

Comment:
I think you can do `.into()` instead of `as i32`

Reply:
Try it and you'll see why I can't